### PR TITLE
Specialise `any` and `all` for tuples

### LINKED ISSUE #40283: Allocation in iterating over a ProductIterator constructed using mixed iterator types
```julia
julia> itp = Iterators.product(1:1:4, 1:5)
Base.Iterators.ProductIterator{Tuple{StepRange{Int64, Int64}, UnitRange{Int64}}}((1:1:4, 1:5))

julia> @btime iterate($itp)
  61.946 ns (2 allocations: 80 bytes)
((1, 1), ((1, 1), (1, 1)))
```

Ideally this should not allocate. This does not allocate if the types of the iterators are identical

```julia
julia> itp = Iterators.product(1:4, 1:5)
Base.Iterators.ProductIterator{Tuple{UnitRange{Int64}, UnitRange{Int64}}}((1:4, 1:5))

julia> @btime iterate($itp)
  4.260 ns (0 allocations: 0 bytes)
((1, 1), ((1, 1), (1, 1)))

julia> itp = Iterators.product(1:1:4, 1:1:5)
Base.Iterators.ProductIterator{Tuple{StepRange{Int64, Int64}, StepRange{Int64, Int64}}}((1:1:4, 1:1:5))

julia> @btime iterate($itp)
  5.316 ns (0 allocations: 0 bytes)
((1, 1), ((1, 1), (1, 1)))
```

### PR DESCRIPTION
**Pull request by jipolanco**
This avoids a small allocation (and improves performance) when iterating over a `ProductIterator` constructed using mixed iterator types. See #40283 for details.

I couldn't really identify where the allocation comes from, and there's likely an underlying issue to be still fixed. But this does solve the `ProductIterator` issue.

When used with homogeneous iterator types, this has no additional cost compared to the original short-circuiting version, as seen in the example below.

Also note that `Base.map` (used in this PR) was already being used in `isdone(z::Zip)`.

Closes #40283.

Example:

```julia
using BenchmarkTools

# 1. Mixing different iterator types (heterogeneous iterator)
xs = (1:5, 1:0.5:4)
iter = Iterators.product(xs...)

# Julia master:
@btime iterate($iter);  # 37.649 ns (2 allocations: 144 bytes)

# With this PR:
@btime iterate($iter);  # 8.586 ns (0 allocations: 0 bytes)

# 2. Using the same iterator type (homogeneous iterator)
xs = (1:5, 2:4)
iter = Iterators.product(xs...)

# Julia master:
@btime iterate($iter);  # 2.181 ns (0 allocations: 0 bytes)

# With this PR:
@btime iterate($iter);  # 2.124 ns (0 allocations: 0 bytes)
```

### CODE REVIEW

**Comment by N5N3** [comment]
`any(f, iter::Tuple)` use loop-based fallback thus any mixed types would do harm to performance.
Maybe we should force inline `iter` with small length (similar to `Base._tuple_any`)

File: base/iterators.jl [hunk 1]
```diff
@@ -1032,7 +1032,7 @@ _prod_eltype(::Type{I}) where {I<:Tuple} =
 iterate(::ProductIterator{Tuple{}}) = (), true
 iterate(::ProductIterator{Tuple{}}, state) = nothing
 
-@inline isdone(P::ProductIterator) = any(isdone, P.iterators)
+@inline isdone(P::ProductIterator) = any(Base.map(isdone, P.iterators))  # issue #40283
```
Comment (simeonschaub) [inline base/iterators.jl:1035]: We should make sure to keep the sort-circuiting behavior here and not unconditionally evaluate `isdone` for every entry in `P.iterators`. I agree with @N5N3 that there should probably be an optimized version of `any` for this.

Comment (jipolanco) [reply base/iterators.jl:1035]: Got it, in that case we can add a version of `any` for tuples. I guess the same should be done for `all`, which seems to have the same issue.

**Comment by jipolanco** [comment]
Seems reasonable to add an optimised version of `any` (and also `all`?) for tuples.

Here is one possible implementation:

```julia
# any(f, itr) = _any(f, itr, :)  # definition in base/reduce.jl

_any(f, itr::Tuple, ::Colon) = _any_tuple(f, false, itr...)

@inline function _any_tuple(f, anymissing, x, rest...)
    v = f(x)
    if ismissing(v)
        anymissing = true
    elseif v
        return true
    end
    return _any_tuple(f, anymissing, rest...)
end

@inline _any_tuple(f, anymissing) = anymissing ? missing : false
```

Should this be limited to small tuples?

**Comment by N5N3** [comment]
I think it's OK to put these optimization in `reduce.jl` to fix the build error.
Another suggestion: if `eltype(x::Tuple)` is concrete, it seems good to use for loop and Let LLVM unroll for us.
```julia
# If eltype(x) is concrete. Use for loop
any(f, x::NTuple{N}) where {N} = _any(f, x, :)
# Otherwise force unroll to avoid union-split. (Only for small Tuple)
function any(f, x::NTuple{N,Any}) where {N}
    N >= 32 && return _any(f, x, :) # Not tuned, just follow Any32
    _any_tuple(f, false, x...)
end
```

**Comment by jipolanco** [comment]
@NSN3 Thanks for the suggestions! This should be taken care of now.

**Comment by KristofferC** [comment]
@jipolanco Is this good to go?

**Pull request merged**